packages/ses/src/module-load.js

```diff
@@ -176,37 +203,123 @@ const loadWithoutErrorAnnotation = async (
     compartment,
     moduleSpecifier,
     staticModuleRecord,
+    queue,
+    loading,
+    errors,
   );
 };
 
-// `load` asynchronously loads `StaticModuleRecords` and creates a complete
-// graph of `ModuleCompartmentRecords`.
-// The module records refer to each other by a reference to the dependency's
-// compartment and the specifier of the module within its own compartment.
-// This graph is then ready to be synchronously linked and executed.
-export const load = async (
+const memoizedLoadWithErrorAnnotation = async (
   compartmentPrivateFields,
   moduleAliases,
   compartment,
   moduleSpecifier,
+  queue,
+  loading,
+  errors,
 ) => {
-  return promiseCatch(
+  const { name: compartmentName } = weakmapGet(
+    compartmentPrivateFields,
+    compartment,
+  );
+
+  // Prevent data-lock from recursion into branches visited in dependent loads.
+  let compartmentLoading = mapGet(loading, compartment);
+  if (compartmentLoading === undefined) {
+    compartmentLoading = new Map();
+    mapSet(loading, compartment, compartmentLoading);
+  }
+  let moduleLoading = mapGet(compartmentLoading, moduleSpecifier);
+  if (moduleLoading !== undefined) {
+    return moduleLoading;
+  }
+
+  moduleLoading = promiseCatch(
     loadWithoutErrorAnnotation(
       compartmentPrivateFields,
       moduleAliases,
       compartment,
       moduleSpecifier,
+      queue,
+      loading,
+      errors,
     ),
     error => {
-      const { name } = weakmapGet(compartmentPrivateFields, compartment);
       // eslint-disable-next-line @endo/no-polymorphic-call
       assert.note(
         error,
         d`${error.message}, loading ${q(moduleSpecifier)} in compartment ${q(
-          name,
+          compartmentName,
         )}`,
       );
       throw error;
     },
   );
+
+  mapSet(compartmentLoading, moduleSpecifier, moduleLoading);
+
+  return moduleLoading;
+};
+
+/*
+ * `load` asynchronously gathers the `StaticModuleRecord`s for a module and its
+ * transitive dependencies.
+ * The module records refer to each other by a reference to the dependency's
+ * compartment and the specifier of the module within its own compartment.
+ * This graph is then ready to be synchronously linked and executed.
+ */
+export const load = async (
+  compartmentPrivateFields,
+  moduleAliases,
+  compartment,
+  moduleSpecifier,
+) => {
+  const { name: compartmentName } = weakmapGet(
+    compartmentPrivateFields,
+    compartment,
+  );
+
+  /** @type {Set<Promise<undefined>>} */
+  const queue = new Set();
```

Comment:
I'm confused by whether you want this to be a `queue` (in which case, I'd expect it to be an `Array<Promise<undefined>>`), or a `pendingJobs` (which is not necessarily a `queue`).

I'd suggest:
```suggestion
  /** @type {Set<Promise<undefined>>} */
  const pendingJobs = new Set();
```